server/data/data-cassandra/src/main/java/org/apache/james/sieve/cassandra/CassandraSieveRepository.java

```diff
@@ -0,0 +1,240 @@
+ ****************************************************************/
+
+package org.apache.james.sieve.cassandra;
+
+import java.io.InputStream;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import javax.inject.Inject;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.james.sieve.cassandra.model.ScriptContentAndActivation;
+import org.apache.james.sieverepository.api.ScriptSummary;
+import org.apache.james.sieve.cassandra.model.SieveQuota;
+import org.apache.james.sieverepository.api.SieveRepository;
+import org.apache.james.sieverepository.api.exception.DuplicateException;
+import org.apache.james.sieverepository.api.exception.IsActiveException;
+import org.apache.james.sieverepository.api.exception.QuotaExceededException;
+import org.apache.james.sieverepository.api.exception.QuotaNotFoundException;
+import org.apache.james.sieverepository.api.exception.ScriptNotFoundException;
+import org.apache.james.sieverepository.api.exception.StorageException;
+import org.joda.time.DateTime;
+
+public class CassandraSieveRepository implements SieveRepository {
+
+    private final CassandraSieveDAO cassandraSieveDAO;
+
+    @Inject
+    public CassandraSieveRepository(CassandraSieveDAO cassandraSieveDAO) {
+        this.cassandraSieveDAO = cassandraSieveDAO;
+    }
+
+    @Override
+    public DateTime getActivationDateForActiveScript(String user) throws StorageException, ScriptNotFoundException {
+        return cassandraSieveDAO.getActiveScriptActivationDate(user)
+            .join()
+            .orElseThrow(ScriptNotFoundException::new);
+    }
+
+    @Override
+    public void haveSpace(String user, String name, long newSize) throws QuotaExceededException, StorageException {
+        throwOnOverQuota(user, spaceThatWillBeUsedByNewScript(user, name, newSize));
+    }
+
+    private void throwOnOverQuota(String user, CompletableFuture<Long> sizeDifference) throws QuotaExceededException, StorageException {
+        CompletableFuture<Optional<Long>> userQuotaFuture = cassandraSieveDAO.getQuota(user);
+        CompletableFuture<Optional<Long>> globalQuotaFuture = cassandraSieveDAO.getQuota();
+        CompletableFuture<Long> spaceUsedFuture = cassandraSieveDAO.spaceUsedBy(user);
+
+        new SieveQuota(spaceUsedFuture.join(), limitToUse(userQuotaFuture, globalQuotaFuture)).checkOverQuotaUponModification(sizeDifference.join());
+    }
+
+    public CompletableFuture<Long> spaceThatWillBeUsedByNewScript(String user, String name, long scriptSize) {
+        return cassandraSieveDAO.getScriptSize(user, name)
+            .thenApply(sizeOfStoredScript -> scriptSize - sizeOfStoredScript.orElse(0L));
+    }
+
+    private Optional<Long> limitToUse(CompletableFuture<Optional<Long>> userQuota, CompletableFuture<Optional<Long>> globalQuota) {
+        if (userQuota.join().isPresent()) {
+            return userQuota.join();
+        }
+        return globalQuota.join();
```

Comment:
method calls are not lazy so the second join would be executed in every case and it's not needed